### DISCUSSION TITLE
OUT column handling in v2

### DISCUSSION THREAD

**Discussion by thomcuddihy** [LAPKB/Pmetrics#144] [Ideas]
Issue: legacy v1 datasets have used '.' in the CSV OUT column as per original specifications, but this is no longer appropriate for v2.

Proposal: Keep existing implementation of non-numeric testing in `PMcheck() errcheck()`, but add functionality to `PMcheck() errfix()` to replace '.' in OUT column only with a static integer.

Question: I have not inspected the code base entirely. Would using a static integer (like -99 when EVID is missing) affect the algorithms?

Pseudo-code Usage:
```
> NPrun(model = "model.txt", data = "data.csv")
ERR: non-numerical column
> fixedData <- PMcheck(data = "data.csv", fix = T)
ERR: non-numerical column
Fixed OUT column
> NPrun(model = "model.txt", data = fixedData)
Running...
```

Happy to implement.

**Comment by mnneely** [comment]
Hi Thom,

In legacy Pmetrics, if EVID = 0, there had to be a numeric entry. Users could use -99 to indicated that the data were missing. For the other supported EVID (1 and 4, dose events), OUT was not relevant, so a "." was used as a placeholder.

In v2 Pmetrics, the data format has become more flexible to the user (although is standardized under the hood to the legacy format). Now, if EVID is missing (which was not possible in legacy Pmetrics), any row with a numeric value in the OUT column is assumed to be EVID = 0. If the user wishes to specify an observation event but with missing data, then as before, -99 is the value for the OUT column in that row event. There should then be "." in the DOSE column. 

Likewise, for DOSE events, the DOSE column should have a numeric value, and the OUT column should have ".". 

There should never be an event with DOSE and OUT both set to ".". Is that the case you are defining to update PMcheck?

**Comment by thomcuddihy** [comment]
Ah, please forgive me, as I'm not personally familiar with the software or the various datasets I was given. As you may have guessed from the PR org name FordeGenomics, I deal more with genetics than pharmacokinetics, and was asked to have a look at this by the Roberts group at UQCCR.

Inspecting their data further, the issue isn't to do with OUT column specifically, but how `readr::read_delim` builds the table from the CSV. Here column 5 is DOSE and column 9 is OUT, with both entries being '.' in the CSV:
```
> data2[, 5][5]
[1] NA
> data2[, 9][1]
[1] NA
> is.numeric(data2[, 5][5])
[1] TRUE
> is.numeric(data2[, 9][1])
[1] FALSE
> typeof(data2[, 5][5])
[1] "double"
> typeof(data2[, 9][1])
[1] "character"
```
On checking their inputs, one of the commented lines was modified to include a note of why it was commented out, and `readr` used that to cast the column type as character:
```
Rows: 655 Columns: 28
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────────Delimiter: ","
chr  (2): #ID, OUT
dbl (20): EVID, TIME, DUR, DOSE, INPUT, OUTEQ, ICU, MALE, AGE, WT, HT, ALB, ...
lgl  (6): ADDL, II, C0, C1, C2, C3
```
While this is most definitely a user error, I would suggest casting all `NA` values to double before returning as this was a bit annoying to track down, since the offending commented rows are removed by `PMreadMatrix()` and therefore not in the `errors.xlsx` spreadsheet.

Happy to implement if you think it is appropriate, otherwise I'll just write a script to fix their CSVs

**Reply by mhovd** [reply]
Do you experience the same issue when you use `PM_data$new("data.csv")`?
